aptos-move/aptos-vm/src/aptos_vm_impl.rs

```diff
@@ -146,22 +146,42 @@ impl AptosVMImpl {
         })
     }
 
-    pub fn check_gas(
+    pub fn check_gas<S: MoveResolverExt>(
         &self,
+        storage: &S,
         txn_data: &TransactionMetadata,
         log_context: &AdapterLogSchema,
     ) -> Result<(), VMStatus> {
         let txn_gas_params = &self.get_gas_parameters(log_context)?.txn;
         let raw_bytes_len = txn_data.transaction_size;
         // The transaction is too large.
         if txn_data.transaction_size > txn_gas_params.max_transaction_size_in_bytes {
-            warn!(
-                *log_context,
-                "[VM] Transaction size too big {} (max {})",
-                raw_bytes_len,
-                txn_gas_params.max_transaction_size_in_bytes,
-            );
-            return Err(VMStatus::Error(StatusCode::EXCEEDED_MAX_TRANSACTION_SIZE));
+            let data =
+                storage.get_resource(&CORE_CODE_ADDRESS, &ApprovedExecutionHashes::struct_tag());
+
+            let valid = if let Ok(Some(data)) = data {
+                let approved_execution_hashes =
+                    bcs::from_bytes::<ApprovedExecutionHashes>(&data).ok();
+                approved_execution_hashes
+                    .map(|aeh| {
+                        aeh.entries
+                            .into_iter()
+                            .any(|value| value.1 == txn_data.script_hash)
```

Comment:
For readability, it's probably better to do `|(_, hash)| hash == txn_data.script_hash`

Reply:
Ha